src/SB/Core/x/xParEmitter.cpp

```diff
@@ -1,3 +1,828 @@
+    }
+    else
+    {
+        bEmitter->parSys = NULL;
+    }
+    t = prop;
+    bEmitter->attachTo = 0;
+    bEmitter->emit_volume = 0;
+    for (i = 0; i < 0xE; i += 1)
+    {
+        if (t->param[1] == t->param[2])
+        {
+            t->param[3] = 0U;
+        }
+        else
+        {
+            t->param[3] = (F32)xParInterpConvertInterpMode((xParInterp*)(t->param + 1));
+        }
+        t += 0x14;
+    }
+    bEmitter->rate_mode = (U8)prop->param[3];
+    if (((U8)bEmitter->eventFunc == 2) && (prop->param[1] == prop->param[2]))
+    {
+        bEmitter->rate_mode = 0;
+    }
+    if (bEmitter->rate_mode == 2)
+    {
+        (bEmitter->prop->life).order();
+    }
+    if (bEmitter->rate_mode == 2)
+    {
+        temp_rate = ((prop->param[1] - prop->param[2]) * xurand()) + prop->param[2];
+    }
+    else
+    {
+        temp_rate = prop->param[1];
+    }
+    bEmitter->rate = temp_rate;
+    bEmitter->rate_time = 0.0f;
+    bEmitter->rate_fraction = 0.0f;
+    bEmitter->rate_fraction_cull = 0.0f;
+    if (pea->emit_flags & 2)
+    {
+        temp_oocull = 1.0f / pea->cull_dist_sqr;
+    }
+    else
+    {
+        temp_oocull = 0.0f;
+    }
+    bEmitter->oocull_distance_sqr = temp_oocull;
+    bEmitter->emit_volume = 0;
+    if ((pea->emit_type == 0xE) || (pea->emit_type == 0xD))
+    {
+        vec_length = (pea->e_circle.dir).length2();
+        if (vec_length < 0.001f)
+        {
+            pea->e_circle.dir = 0;
+        }
+        else
+        {
+            pea->e_circle.dir = 1.0f / xsqrt(vec_length);
+        }
+    }
+    bEmitter->last_attach_loc += pea->e_circle.dir;
+}
+
+void xParEmitterSetup(xParEmitter* t)
+{
+    xParEmitterAsset* pea;
+    xBase* temp_base;
+
+    if (t->parSys != (xParSys*)0x0)
+    {
+        t->group = *(xParGroup**)(t->parSys + 0x20);
+    }
+    if ((t->tasset->attachToID != 0) && (t->tasset->emit_type == '\t'))
+    {
+        temp_base = zSceneFindObject(t->tasset->attachToID);
+        iModelTagSetup(&t->tag, *(RpAtomic**)(*(S32*)&temp_base[2].baseType + 0x10),
+                       *(F32*)&t->tasset->e_entbound, (t->tasset->e_entbound).expand,
+                       (t->tasset->e_entbound).deflection);
+    }
+    temp_base = (xBase*)0x0;
+    if (t->tasset->attachToID != 0)
+    {
+        temp_base = zSceneFindObject(t->tasset->attachToID);
+    }
+    t->attachTo = temp_base;
+    if (((t->tasset->emit_type != '\x10') && (t->tasset->emit_type != '\x0F')) &&
+        (t->tasset->emit_type == '\x06'))
+    {
+        temp_base = zSceneFindObject(t->tasset->attachToID);
+        t->emit_volume = temp_base;
+    }
+    add_tweaks(*t);
+    return;
+}
+
+void xParEmitterReset(xParEmitter* t)
+{
+    xBaseReset((xBase*)t, t->tasset);
+    t->emit_flags = t->tasset->emit_flags;
+}
+
+S32 xParEmitterEventCB(xBase* to1, xBase* to2, U32 toEvent, F32* unused1, xBase* unused2)
+{
+    xParEmitterCustomSettings sp8;
+
+    switch ((S32)toEvent)
+    {
+    case 0xA:
```

Comment:
Please replace ALL of these with the eEvent enums in xEvent if they match. Those are used throughout the codebase.